### ISSUE TITLE
[Feature Request] Custom versions instead of latest at the time of filterResults/target

### ISSUE THREAD

**Issue by 5punk** [raineorshine/npm-check-updates#1314]
- [x] I have searched for [similar issues](https://github.com/raineorshine/npm-check-updates/issues)
- [x] I am using the latest version of `npm-check-updates`
- [x] I am using `node >= 14.14`

---

## Current Behavior

We are able to use this package as intended and are getting the desired behavior. However, this library enables always enforces the latest package version for each of the "monitored" packages. It is common in enterprises, for application security teams "bless" versions of a package before they can be consumed. We maintain a list of packageNames and blessed versions but are unable to override/overwrite the default `upgradeVersion` found.

Closest alternatives:
- The function `filterResults` today only accepts a boolean override
- The function `target` today only accepts a tag: `latest, newest, greatest, minor, patch, semver, @[tag]`

```
Target fn Error: 
Unsupported target "4.0.2" using yarn
Supported version targets are: latest, newest, greatest, minor, patch, and tags (e.g. @next)
```

## Proposal

I wanted to propose or inquire if the behavior of `filterResults` or `target` can be extended to do the following:

```javascript
{
  filterResults: (packageName, { currentVersion, upgradedVersion }) => {
    const approvedVersions = {
       [packageName]: "1.2.3"
    };
    // allow mutation of upgradedVersion
    upgradedVersion = approvedVersion[packageName];
    return true;
  }
}

```

OR

**Cleaner**

```javascript
{
  target: (dependencyName, [{ semver, version, operator, major, minor, patch, release, build }]) => {
    const approvedVersions = {
       [dependencyName]: "1.2.3"
    };
    // allow return of semver TYPE as valid target
    return SemVer(approvedVersion[dependencyName]);
  }
}

```

Willing to contribute a PR.

**Comment by raineorshine** [comment]
Hi, thanks for the suggestion.

One way to accomplish the desired functionality is to run a custom `--registry` that only exposes the blessed versions. Or, if you have a static list of blessed versions you can use `--packageManager staticRegistry`. You could even fetch a dynamic version list, save it to disk, and then use it as a static registry before each `npm-check-updates` call. 

Would that work? If not, I'm happy to discuss the custom target function approach.

**Comment by 5punk** [comment]
Oh, how convenient!

Apologies, I missed this 😓 
I'll try it out and report back. Looks very promising.

**Comment by raineorshine** [comment]
No problem! npm-check-updates has a lot of features, so it's easy to miss.

**Comment by 5punk** [comment]
Worked great, bamboozled by the follow-up installation command.

![image](https://github.com/raineorshine/npm-check-updates/assets/8335261/1994e236-a4cf-48ad-b100-5a97ebb221a9)

If I try to use the static registry file path with `yarn`, I get the following error since it's expecting a URL and not a file path:

`ncu -i -p yarn -r ./path/file.json`

![image](https://github.com/raineorshine/npm-check-updates/assets/8335261/bf3261a3-f8a6-4bd2-b4ff-67df34d83978)

**Comment by 5punk** [comment]
`staticRegistry install` should never be a thing, sounds like a bug right?

How about we remove `staticRegistry` as a package manager. I tried to pass in a `file://` URL to see if I can override this, but I get a `TypeError: Only HTTP(S) protocols are supported` from mini-pass-fetch.

I wonder if the right thing to do is to allow path names to json files as a valid registry value instead of only URLs.

**Comment by raineorshine** [comment]
> Worked great, bamboozled by the follow-up installation command.

Oops! 🐞

> `staticRegistry install` should never be a thing, sounds like a bug right?

It should never be a thing 😶‍🌫️

> I wonder if the right thing to do is remove `staticRegistry` as a package manager and allow path names to json files as a valid registry value instead of only URLs OR allow for a new param `staticRegistryPackageManager` to be used in conjunction.

I opted to autodetect the package manager when `--staticRegistry` is specified.

Also added static registry support for `--registry {URL}`. Without `--staticRegistry`, the registry is expected to be a full npm registry. With `--staticRegistry`, it should be a simple JSON file.

Published to `v16.12.0`

**Comment by 5punk** [comment]
The autodetect picked up `npm` rather than `yarn` since there is no lock file at the root of the cwd() -- which will be the case with monorepos.... 

I don't want to push it, if you feel ncu should not support this usecase for monorepos I can understand. 

I feel the cleanest and most explicit option for your consideration would be to just allow `staticRegistryPackageManager` as an option that can be checked/ingested [here](https://github.com/raineorshine/npm-check-updates/blob/ad07158adb12f16c6f2a38f5d22d335a199e74fc/src/lib/initOptions.ts#L146-L149), it can fallback/default to auto detect.

**Comment by raineorshine** [comment]
Ah, good point.

`staticRegistryPackageManager` makes sense, although I think it is kind of covering up my poor API design. `staticRegistryPackageManager` overrides `packageManager` which overrides the original package manager... 😅

Rethinking the static registry API completely, how about this:

- Deprecate `--packageManager staticRegistry`
- Add option `--registryType [npm | json]`. Default: `npm`.
- Set `--registryType json` and `--registry` together for static registry behavior.
- `--packageManager` is preserved for use in conjunction with static registries.

**Comment by 5punk** [comment]
> is kind of covering up my poor API design.

Not at all! I'm a fan, you've just been fielding so many feature requests and contribs - it's easy to lose control.

> `--registryType [npm | json]` and --packageManager is preserved for use in conjunction

I didn't want to assume as much freedom - This is so much more cleaner! Keeping registry type explicit (api/static OR npm/json) AND `packageManager` pure and without any overlap with overriding behavior.

**Comment by raineorshine** [comment]
Great, thanks so much for your feedback.

**Comment by raineorshine** [comment]
New `--registryType` api published to `v16.12.2`

## registryType

Usage:

    ncu --registryType [type]

Specify whether --registry refers to a full npm registry or a simple JSON file.

<table>
  <tr><td>npm</td><td>Default npm registry</td></tr>
  <tr><td>json</td><td>Checks versions from a file or url to a simple JSON registry. Must include the `--registry` option.

Example:

    // local file
    $ ncu --registryType json --registry ./registry.json

    // url
    $ ncu --registryType json --registry https://api.mydomain/registry.json

    // you can omit --registryType when the registry ends in .json
    $ ncu --registry ./registry.json
    $ ncu --registry https://api.mydomain/registry.json

registry.json:

    {
      "prettier": "2.7.1",
      "typescript": "4.7.4"
    }

</td></tr>
</table>
